File: tools/wave/src/testing/event-dispatcher.js
const STATUS_EVENT = "status";
const RESUME_EVENT = "resume";
const TEST_COMPLETED_EVENT = "test_completed";

class EventDispatcher {
  constructor() {
    this._clients = {};
  }

  addSessionClient(client) {
    const token = client.getSessionToken();
    if (!this._clients[token]) this._clients[token] = [];
    this._clients[token].push(client);
  }

  removeSessionClient(clientToDelete) {
    if (!clientToDelete) return;
    const token = clientToDelete.getSessionToken();
    if (!this._clients[token]) return;
    this._clients[token].splice(
      this._clients[token].findIndex(client => client === clientToDelete),
      1
    );
    if (this._clients[token].length === 0) delete this._clients[token];
  }

  dispatchEvent({ token, type, data }) {
    if (!this._clients[token]) return;
    const event = { type, data };
    this._clients[token].forEach(client =>
      client.sendMessage(JSON.stringify(event))
    );
  }
}

EventDispatcher.STATUS_EVENT = STATUS_EVENT;
EventDispatcher.TEST_COMPLETED_EVENT = TEST_COMPLETED_EVENT;
EventDispatcher.RESUME_EVENT = RESUME_EVENT;

module.exports = EventDispatcher;

File: tools/wave/testing/event_dispatcher.py
from __future__ import unicode_literals
import uuid


STATUS_EVENT = "status"
RESUME_EVENT = "resume"
TEST_COMPLETED_EVENT = "test_completed"

DEVICES = "devices"
DEVICE_ADDED_EVENT = "device_added"
DEVICE_REMOVED_EVENT = "device_removed"

class EventDispatcher(object):
    def __init__(self):
        self._listeners = {}

    def add_event_listener(self, listener):
        token = listener.dispatcher_token
        if token not in self._listeners:
            self._listeners[token] = []
        self._listeners[token].append(listener)
        listener.token = str(uuid.uuid1())
        return listener.token

    def remove_event_listener(self, listener_token):
        if listener_token is None:
            return

        for dispatcher_token in self._listeners:
            for listener in self._listeners[dispatcher_token]:
                if listener.token == listener_token:
                    self._listeners[dispatcher_token].remove(listener)
                    if len(self._listeners[dispatcher_token]) == 0:
                        del self._listeners[dispatcher_token]
                    return

    def dispatch_event(self, dispatcher_token, event_type, data=None):
        if dispatcher_token not in self._listeners:
            return
        event = {
            "type": event_type,
            "data": data
        }

        for listener in self._listeners[dispatcher_token]:
            listener.send_message(event)
